File: bn_parser.c
#include "bn_parser.h"
#include "bn_utils.h"
#include "lexer.h"

#include <stdlib.h>

struct token tk;

struct non_term {
	const char *name;
	struct symbol_list *curr_prod;
} curr_nt;

void add_prod()
{
	// TODO: move this to bn_utils
	struct prod_list *new_prod = malloc(sizeof(struct prod_list));
	new_prod->next = look_up(curr_nt.name)->prods;
	new_prod->prod = malloc(sizeof(struct symbol_list));
	curr_nt.curr_prod = new_prod->prod;
	curr_nt.curr_prod->rem_syms = 0;
}

void parse_prods()
{
	// TODO
	add_prod();
	while (next_token(&tk)) {
		if (tk.type == TK_BAR) {
			add_prod();
			continue;
		}
		/* parse next symbol in prod */
		if (tk.type == TK_LESS) {	/* parse non term */
			next_token(&tk);
			if (tk.type != TK_ID)
				panic("expected non terminal symbol");
			// TODO: char *nt = strdup(tk.str_val);
			next_token(&tk);
			if (tk.type != TK_GRT)
				panic("expected '>'");
			next_token(&tk);
			if (tk.type == TK_COLN) {	/* we are in a new definition "<nt> :=" */
				next_token(&tk);
				if (tk.type != TK_ASSIGN)
					panic("expected '='");
				// TODO: curr_nt.name = create_entry(nt)->key;
				// TODO: new_def(nt);
				continue;
			}
			/* store the non term in the current prod */
			/* XXX: the prod is stored in reverse order */
			struct symbol_list *new_sym = malloc(sizeof(struct symbol_list));
			new_sym->is_term = 0;
			// TODO: new_sym->nt_name = nt;
			new_sym->rem_syms = curr_nt.curr_prod->rem_syms + 1;
			new_sym->next = curr_nt.curr_prod;
			curr_nt.curr_prod = new_sym;
		}
		if (tk.type == TK_STR) {	/* parse literal */
			/* store the term in the current prod */
			/* XXX: the prod is stored in reverse order */
			struct symbol_list *new_sym = malloc(sizeof(struct symbol_list));
			new_sym->is_term = 1;
			new_sym->term_val = TK_LPAR;	/* TODO: implement this */
			new_sym->rem_syms = curr_nt.curr_prod->rem_syms + 1;
			new_sym->next = curr_nt.curr_prod;
			curr_nt.curr_prod = new_sym;
		}
	}
}

void parse_bn()
{
	// TODO
}

File: bn_parser.py
import sys
from collections import namedtuple

# XXX: these are defined in "lexer.h"
TK_STR = 149
TK_ID = 150

grammar = dict()
Token = namedtuple("Token", ["type", "val"])
tk = Token(0, 0)
curr_nt = ""
curr_prod = list()
tk_n = 0

def print_token(tk):
    print(tk.type, end='\t')
    if tk.type == TK_ID or tk.type == TK_STR:
        print(tk.val)
        return
    print(chr(tk.type))

def tk_gen():
    global tk_n
    for tk in sys.stdin.readlines():
        t, v = tk.split('\t', 1)
        tk_n += 1
        yield (int(t), v.rstrip())

tokens = tk_gen()

def expected(exp):
    raise Exception(
        f"on token {tk_n} \"{chr(tk.type)}\": "
        f"expected {exp}"
    )

def next_token():
    global tk
    tk = Token(*tokens.__next__())

def skip_chars(chars):
    for char in chars:
        if tk.type != ord(char):
            expected(chars)
        next_token()

def parse_prods():
    global tk
    global curr_nt
    global curr_prod

    while True:
        if tk.type == ord('|'): # new prod for curr_nt
            grammar[curr_nt].append(curr_prod.copy())
            curr_prod = list()
            next_token()
            continue

        if tk.type == ord('<'): # parse <symbol>
            next_token()
            if tk.type != TK_ID:
                expected("\"symbol\"")
            nt = tk.val
            next_token()

            if tk.type != ord('>'):
                expected("'>'")
            next_token()

            if tk.type == ord(':'): # we are in a new def
                skip_chars("::=")
                # save curr_prod for this nonterm
                grammar[curr_nt].append(curr_prod.copy())

                curr_nt = nt # start prod for newly defined nonterm
                if grammar.get(curr_nt) == None:
                    grammar[curr_nt] = list()
                curr_prod = list()
                continue

            curr_prod.append(nt) # add symbol to current prod

        if tk.type == TK_STR: # parse literal symbol
            curr_prod.append(tk.val)
            next_token()
            continue

def parse_bn():
    global tk
    global curr_nt
    global curr_prod

    next_token()
    if tk.type != ord('<'):
        expected("\"<symbol>\"")
    next_token()
    if tk.type != TK_ID:
        expected("\"symbol\"")

    curr_nt = tk.val
    next_token()

    if tk.type != ord('>'):
        expected("'>'")
    next_token()

    skip_chars("::=")

    if grammar.get(curr_nt) == None:
        grammar[curr_nt] = list()
    curr_prod = list()

    try:
        parse_prods()
    except StopIteration:
        return

    

if __name__ == "__main__":
    parse_bn()
    for nt, prods in grammar.items():
        print("<", nt, ">", sep="")
        for prod in prods:
            print("\t", prod, sep="")
